### ISSUE TITLE
ibm_db_dbi Exceptions automatically suppressed when using WITH ... AS ... -syntax

### ISSUE THREAD

**Issue by SirHooke** [ibmdb/python-ibmdb#553]
Hello all,

I just noticed that, when using the WITH syntax, ALL Exceptions appear te be ignored by default, but only AFTER an 'execute' has been run on the cursor.
For illustration, please see the attached [(semi-pseudo code) script](https://github.com/ibmdb/python-ibmdb/files/5275355/ibm_db_dbi_issue.txt) .

My configuration: Python 3.7.3, ibm-db 3.0.2

To summarize: I have a script that executes a query on a (usually DB2) server, and then inserts the result into another (usually MySQL) server.

### The Function
A simple example of the function:

```python
def transfer(src_conn, tgt_conn, qry, insert_sql):
    """
        src_conn: the Connection on which to execute the query
        tgt_conn: the Connection on which to store the result of the query
        qry: the query itself
    """
    print("begin")
    with src_conn.cursor() as src_cur, tgt_conn.cursor() as tgt_cur:
        print("query source")
        src_cur.execute(qry)
        print("start fetch")
        result = src_cur.fetchmany()
        while result:
            print("inserting")
            tgt_cur.executemany(insert_sql, result)
            print("fetching")
            result = src_cur.fetchmany()
    print("end")
```
In my case, `src_conn` is the DB2 connection.

### Normal operation
When no errors occur, output should resemble:

```
begin
query source
start fetch
inserting
fetching
inserting
fetching
inserting
fetching
end
```
#### Before 'execute'
When an error occurs before the `src_cur.execute()` statement, the script throws an exception and stops. This is the desired behaviour.
Sample output:
```
begin
Traceback (most recent call last):
  File "tmp.py", line 121, in <module>
    transfer(db2_connection, mysql_connection_2, qry) 
  File "tmp.py", line 41, in transfer_1
    raise CustomException('Exception before query')
__main__.CustomException: Exception before query
```
#### After 'execute'
This is were the issue exists.
If an exception, **any** exception occurs within the `with .... as ....` block, after `execute()` has been called on the DB2-cursor, the error is silently ignored:
Sample output:
```
begin
query source
end
```
When, on the other hand, `src_cur` and `src_conn` are not DB2, but for example MySQL, the exception is raised correctly.
Sample output:
```
begin
query source
Traceback (most recent call last):
  File "tmp.py", line 121, in <module>
   transfer(mysql_connection, mysql_connection_2, qry) 
  File "tmp.py", line 65, in transfer_2
    raise CustomException('Exception after query')
__main__.CustomException: Exception after query
```

This would lead me to believe that this is a bug within ibm_db_dbi.

**Comment by imavo** [comment]
Can you confirm that if you __omit__ the 'WITH ...' syntax, that the symptom does __NOT__ happen?  

Also, to simplify, if both the source and the target are in different Db2 databases does the symptom happen in the same manner?

**Comment by SirHooke** [comment]
Just did the test without the context manager:
```python
def transfer_man(src_conn, tgt_conn, qry, insert_sql):
    print("begin")
    src_cur = src_conn.cursor()
    tgt_cur = tgt_conn.cursor()
    src_cur.arraysize = 5
    print("query source")
    src_cur.execute(qry)
    print("start fetch")
    result = src_cur.fetchmany()
    while result:
        print("inserting")
        tgt_cur.executemany(insert_sql, result)
        print("fetching")
        result = src_cur.fetchmany()
    src_cur.close()
    tgt_cur.close()
    print("end")
```
Output without errors is exactly the same as the original.
Modified to produce error after `src_cur.execute()`:
```python
def transfer_man_2(src_conn, tgt_conn, qry, insert_sql):
    print("begin")
    src_cur = src_conn.cursor()
    tgt_cur = tgt_conn.cursor()
    src_cur.arraysize = 5
    print("query source")
    src_cur.execute(qry)
    raise CustomException('Exception after query')
    print("start fetch")
    result = src_cur.fetchmany()
    while result:
        print("inserting")
        tgt_cur.executemany(insert_sql, result)
        print("fetching")
        result = src_cur.fetchmany()
    src_cur.close()
    tgt_cur.close()
    print("end")
```
Output is as follows:
```
begin
query source
Traceback (most recent call last):
  File "tmp.py", line 157, in <module>
    transfer_man_2(db2_connection, mysql_connection_2, qry, insert_sql)
  File "tmp.py", line 99, in transfer_man_2
    raise CustomException('Exception after query')
__main__.CustomException: Exception after query
```

This confirms the problem does not exist when not using a WITH syntax.

Unfortunately I don't have a db2 at my disposal where I have write privileges, so I can't verify your second point.

**Comment by imavo** [comment]
In your sample, what is the value of `insert_sql` ? I will try to reproduce with source and target(s) being Db2 databases.

**Comment by SirHooke** [comment]
In my case it is an INSERT statement into a MySQL table, it is dependant on `qry`:
i.e.:

- qry: `SELECT field1, field2, field3 FROM MYDB.TABLE`
- insert_sql: `INSERT INTO MYDB.TABLE2 (field1, field2, field3) VALUES (%s, %s, %s)`

**Comment by imavo** [comment]
Yes, Db2 parameter markers are ?.  I'm confused by your sample, as it seems to define the transfer functions with three argument, but invoke them with four arguments.

**Comment by SirHooke** [comment]
The example script was just something that I made quickly, it did not contain all the details of the test I ran, sorry about that!
I have tried to make it a bit more clear in [this version](https://github.com/ibmdb/python-ibmdb/files/5276702/ibm_db_dbi_issue.txt).

**Comment by imavo** [comment]
I reproduced the issue when source & target are both Db2 databases. Specifically when the transfer_2 and transfer_3 functions run, no exceptions are reported, despite the explicit raise in both cases.
Needs investigating.

**Comment by amukherjee28** [comment]
@imavo can you please provide your repro program/sample. This will help me to debug the issue and investigate.
Since you have already been able to reproduce let me use your program to investigate.
Thanks.

**Comment by imavo** [comment]
@amukherjee28  Attached (i hope).
[dbi_cursors3.txt](https://github.com/ibmdb/python-ibmdb/files/5277706/dbi_cursors3.txt)

**Comment by amukherjee28** [comment]
Hi,

Based on my investigation on the above issue, here are my findings:

CASE 1 :

when using WITH clause no exception is shown and all of the exceptions are suppressed.
In case there is no error programs flow is as expected.

CASE 2:

When WITH clause is removed and normal assignment is made, in case there are error, The error message gets printed.
In case there is no error program flow is as expected.

Please check the attached file which explains both the cases.

[issue553.txt](https://github.com/ibmdb/python-ibmdb/files/5278193/issue553.txt)

Hence I will advice you change the code while I investigate what goes wrong when WITH is used.

Thanks.

**Comment by amukherjee28** [comment]
Hi @SirHooke 
Can you close on this issue. This was application code written in different way.

**Comment by bkhl** [comment]
I just ran into this as well, and it's quite an annoying papercut. @amukherjee28 have you been working on this in some separate issue already or should I create a new one?

**Comment by amukherjee28** [comment]
Hi @bkhl 

Since the issue is closed can you please open a new one with the issue you are facing.
We can investigate.

Thanks.
